src/nodes.coffee

```diff
@@ -57,6 +60,34 @@ exports.Base = class Base
   compile: (o, lvl) ->
     fragmentsToText @compileToFragments o, lvl
 
+  # Occasionally a node is compiled multiple times, for example to get the name
+  # of a variable to add to scope tracking. When we know that a “premature”
+  # compilation won’t result in comments being output, set those comments aside
+  # so that they’re preserved for a later `compile` call that will result in
+  # the comments being included in the output.
+  compileWithoutComments: (o, lvl, method = 'compile') ->
+    if @comments
+      @ignoreTheseCommentsTemporarily = @comments
```

Comment:
The `comments` property already exists on the node before its `compile` method is called. (It’s created by `addDataToNode`, called from `parser.js`.) If I tried to use a flag, that would mean updating the nodes that use `compileWithoutComments` so that each such node’s `compile` or `compileNode` methods reimplemented the logic you see here. There would be a lot more duplication.

The point of this method is that sometimes nodes compile children while discarding side effects. Search for places where I call `compileWithoutComments` and you’ll see. This happens e.g. to get the name of a variable, but then that name is used as a string and the `Value` or `IdentifierLiteral` node never gets properly compiled itself later on. In many such cases I move the comments of a node’s children to the node itself in the node’s constructor (see the constructor for `For`, for example) to avoid the issue; but in some cases it’s unavoidable. (Function parameters that are splats are a particularly nasty case, and I couldn’t figure out how to use this method and instead had to implement a variation of it inside `Code`.) You could argue that cases where `compileWithoutComments` is necessary are cases that are poorly implemented in the first place, and we should refactor those nodes’ `compile` or `compileToFragments` methods . . . and you’d probably be right, at least in most cases. But this PR is big enough as it is, without taking on that additional scope.